crates/shared/src/event_handling.rs

```diff
@@ -91,47 +91,52 @@ where
     }
 
     pub fn last_handled_block(&self) -> Option<BlockNumberHash> {
-        self.last_handled_block
+        self.last_handled_blocks.last().cloned()
     }
 
-    async fn event_block_range(&self) -> Result<RangeInclusive<BlockNumber>> {
-        // Instead of using only the most recent event block from the db we also store the last
-        // handled block in self so that during long times of no events we do not query needlessly
-        // large block ranges.
-        let (last_handled_block_number, last_handled_block_hash) = match self.last_handled_block {
-            Some(block) => block,
-            None => self.store.last_event_block().await?,
-        };
-        let current_block = self.block_retriever.current_block().await?;
-        let current_block_number = current_block
-            .number
-            .context("missing block number")?
-            .as_u64();
-        let from_block = if Some(current_block.parent_hash) == last_handled_block_hash {
-            // no reorg
-            (last_handled_block_number, last_handled_block_hash)
-        } else {
-            (
-                last_handled_block_number.saturating_sub(MAX_REORG_BLOCK_COUNT),
-                None,
+    async fn event_block_range(
+        &self,
+    ) -> Result<(RangeInclusive<BlockNumber>, Vec<BlockNumberHash>)> {
+        let current_blocks = self
+            .block_retriever
+            .block_history(
+                BlockId::Number(Web3BlockNumber::Latest),
+                MAX_REORG_BLOCK_COUNT,
             )
+            .await?;
+        let handled_blocks = if self.last_handled_blocks.is_empty() {
+            let last_handled_block = self.store.last_event_block().await?;
+            self.block_retriever
```

Comment:
@nlordell 
When retrieving handled_blocks, it could happen that the `last_handled_block` (with it's block number and block hash) is actually uncled block at this moment (meaning reorg happened between restarts). This would mean I need [this](https://github.com/cowprotocol/services/issues/478#issuecomment-1230253043) smart contract to be able to give me last 25 blocks (numbers and hashes) based on a given block hash (not given a block number)?

Reply:
I don't think you need it. You could just make a batched request to the node to get block hashes at different depths right?

Reply:
Hm, I can do it to get the current blockchain head and it's history, but at the same time I need a block at specific block height(number) that potentially, at this moment, is uncled block. 
That said, when you ask for a block by block number, you always get the block from canonical chain and not uncled block/s at the same height.

Reply:
Not relevant question anymore.